pkg/ring/ring.go

```diff
@@ -461,6 +462,101 @@ func (r *Ring) Subring(key uint32, n int) ReadRing {
+// number of zones and the returned subring will contain the same number of
+// instances per zone as far as there are enough registered instances in the ring.
+//
+// The algorithm used to build the subring is a shuffle sharder based on probabilistic
+// hashing. We treat each zone as a separate ring and pick N unique replicas from each
+// zone, walking the ring starting from random but predictable numbers. The random
+// generator is initialised with a seed based on the provided identifier.
+//
+// This implementation guarantees:
+// - Stability: given the same ring, two invocations returns the same result.
+// - Consistency: adding/removing 1 instance from the ring generates a resulting
+//   subring with no more then 1 difference.
+// - Shuffling: probabilistically, for a large enough cluster each identifier gets
+//   a different set of instances, with a reduced number of overlapping instances
+//   between two identifiers.
+func (r *Ring) ShuffleShard(identifier string, size int) ReadRing {
+	// Nothing to do if the shard size is not smaller then the actual ring.
+	if size <= 0 || r.IngesterCount() <= size {
+		return r
+	}
+
+	// Use the identifier to compute an hash we'll use to seed the random.
+	hasher := md5.New()
+	hasher.Write([]byte(identifier)) // nolint:errcheck
+	checksum := hasher.Sum(nil)
+
+	// Generate the seed based on the first 64 bits of the checksum.
+	seed := int64(binary.BigEndian.Uint64(checksum))
+
+	// Initialise the random generator used to select instances in the ring.
+	random := rand.New(rand.NewSource(seed))
+
+	r.mtx.RLock()
+
+	// We expect the shard size to be divisible by the number of zones, in order to
+	// have nodes balanced across zones. If it's not, we do round up.
+	numInstancesPerZone := int(math.Ceil(float64(size) / float64(len(r.ringZones))))
+
+	shard := make(map[string]IngesterDesc, size)
+
+	// We need to iterate zones always in the same order to guarantee stability.
+	for _, zone := range r.ringZones {
+		tokens := r.ringTokensByZone[zone]
+
+		// To select one more instance while guaranteeing the "consistency" property,
+		// we do pick a random value from the generator and resolve uniqueness collisions
+		// (if any) continuing walking the ring.
+		for i := 0; i < numInstancesPerZone; i++ {
+			start := searchToken(tokens, random.Uint32())
+			iterations := 0
+			found := false
+
+			for p := start; iterations < len(tokens); p++ {
+				iterations++
+
+				// Wrap p around in the ring.
+				p %= len(tokens)
+
+				// Ensure we select an unique instance.
+				if _, ok := shard[tokens[p].Ingester]; ok {
+					continue
+				}
+
+				shard[tokens[p].Ingester] = r.ringDesc.Ingesters[tokens[p].Ingester]
+				found = true
+				break
+			}
+
+			// If one more instance has not been found, we can stop looking for
+			// more instances in this zone, because it means the zone has no more
+			// instances which haven't been already selected.
+			if !found {
+				break
+			}
+		}
+	}
+
+	r.mtx.RUnlock()
```

Comment:
Any way we can use `defer`? It'll make the code less brittle to future changes.

Reply:
Sure, done. I put it here to keep it locked for the shortest time, but it's really not an issue, especially considering the upcoming cache.